### ISSUE TITLE
Cannot mount configmap from ro, nodev, noexec, nosuid filesystem in rootless mode

### ISSUE THREAD

**Issue by rpluem-vf** [k3s-io/k3s#7436]
**Environmental Info:**
K3s Version:

k3s version v1.26.3+k3s-30638072-dirty (30638072)
go version go1.19.7

Node(s) CPU architecture, OS, and Version:

Linux localhost.localdomain 4.18.0-425.19.2.el8_7.x86_64 #1 SMP Tue Apr 4 22:38:11 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

Cluster Configuration:

No Cluster. Single server started via `/opt/SP/k3sadm/k3s/bin/k3s server --rootless --disable traefik  --data-dir /var/SP/rancher/k3s` from user systemd.

**Describe the bug:**

Mounting a configMap in rootless mode created on a filesystem that has one or more of the following mount options set to a container fails: `ro`, `nodev`, `noexec`, `nosuid`

**Steps To Reproduce:**

1. Build k3s from source.
2. Follow the prequisites for [rootless](https://docs.k3s.io/advanced#running-rootless-servers-experimental).
3. Start k3s rootless via the the following systemd unit as non privileged user:

    ```systemd
    [Unit]
    Description=k3s (Rootless)

    [Service]
    Environment=PATH=/opt/SP/k3sadm/k3s/bin:/usr/sbin:/usr/bin:/sbin:/bin
    # NOTE: Don't try to run `k3s server --rootless` on a terminal, as it doesn't enable cgroup v2 delegation.
    # If you really need to try it on a terminal, prepend `systemd-run --user -p Delegate=yes --tty` to create a systemd scope.
    ExecStart=/opt/SP/k3sadm/k3s/bin/k3s server --rootless --disable traefik --data-dir /var/SP/rancher/k3s
    ExecReload=/bin/kill -s HUP $MAINPID
    TimeoutSec=0
    RestartSec=2
    Restart=always
    StartLimitBurst=3
    StartLimitInterval=60s
    LimitNOFILE=infinity
    LimitNPROC=infinity
    LimitCORE=infinity
    TasksMax=infinity
    Delegate=yes
    Type=simple
    KillMode=mixed

    [Install]
    WantedBy=default.target
    ```

3. Create a configMap on a filesystem that has one or more of the following mount options set: `ro`, `nodev`, `noexec`, `nosuid`. In my case:

   ```
   /dev/mapper/vg1-varspranchervol on /var/SP/rancher type xfs (rw,nodev,relatime,seclabel,attr2,inode64,logbufs=8,logbsize=32k,noquota)
   ```

   and

   ```
   kubectl create configmap test-config --from-file=/var/SP/rancher/test_mount/Hello.txt
   ```

   ```
   ls -ld /var /var/SP/ /var/SP/rancher/ /var/SP/rancher/test_mount/
   drwxr-xr-x. 21 root   root 4096 Apr 17 10:48 /var
   drwxr-xr-x.  4 root   root   35 Apr 17 11:05 /var/SP/
   drwxr-x---.  4 k3sadm k3s    35 May  3 15:27 /var/SP/rancher/
   drwxr-xr-x.  2 k3sadm k3s    23 Apr 18 09:54 /var/SP/rancher/test_mount/
   ```

   ```
   ls -l /var/SP/rancher/test_mount/Hello.txt 
   -rw-r--r--. 1 k3sadm k3s 6 Apr 18 09:54 /var/SP/rancher/test_mount/Hello.txt
   ```

   ```
   cat /var/SP/rancher/test_mount/Hello.txt 
   Hello
   ```

   `k3sadm` is the user used to start `k3s`.

4. Deploy a simple deployment via `kubectl apply -f dummy_deploy.yml`:

   ```yaml
    ---
    apiVersion: apps/v1
    kind: Deployment
    metadata:
    name: test-deployment
    spec:
    replicas: 1
    selector:
        matchLabels:
        app: test
    template:
        metadata:
        labels:
            app: test
        spec:
        containers:
        - image: docker.io/busybox
            name: test-container
            command: [ "sleep", "86400" ]
            volumeMounts:
            - name: config
            mountPath: /mnt/Hello.txt
            subPath: Hello.txt
        volumes:
        - name: config
            configMap:
            name: test-config
   ```

**Expected behavior:**

Pod starts:

```
kubectl get pod -A
NAMESPACE     NAME                                      READY   STATUS    RESTARTS      AGE
kube-system   local-path-provisioner-76d776f6f9-wnbkm   1/1     Running   3 (32s ago)   39h
default       test-deployment-5f4ff88f68-s5hwd          1/1     Running   0             39h
kube-system   coredns-59b4f5bbd5-7rjwh                  1/1     Running   3 (32s ago)   39h
kube-system   metrics-server-7b67f64457-hpf9r           1/1     Running   3 (32s ago)   39h
```

**Actual behavior:**

Pod does not start:

```
kubectl get pod -A
NAMESPACE     NAME                                      READY   STATUS                       RESTARTS      AGE
default       test-deployment-5f4ff88f68-s5hwd          0/1     CreateContainerConfigError   0             39h
kube-system   coredns-59b4f5bbd5-7rjwh                  1/1     Running                      2 (36s ago)   39h
kube-system   local-path-provisioner-76d776f6f9-wnbkm   1/1     Running                      2 (36s ago)   39h
kube-system   metrics-server-7b67f64457-hpf9r           1/1     Running                      2 (36s ago)   39h
```

**Additional context / logs:**

Logs:

```
May 05 07:30:47 localhost.localdomain k3s[25104]: I0505 07:30:47.042246      28 scope.go:115] "RemoveContainer" containerID="9f59281156eeee969bf99e0e52e8b92824891121cf87a4572761e3006e562a0f"
May 05 07:30:47 localhost.localdomain k3s[25104]: E0505 07:30:47.848086      28 mount_linux.go:231] Mount failed: exit status 32
May 05 07:30:47 localhost.localdomain k3s[25104]: Mounting command: mount
May 05 07:30:47 localhost.localdomain k3s[25104]: Mounting arguments: --no-canonicalize -o bind,remount /proc/28/fd/153 /var/SP/rancher/k3s/agent/kubelet/pods/5a3ea7d9-f077-462b-8fcc-645cef659e3a/volume-subpaths/config/test-container/0
May 05 07:30:47 localhost.localdomain k3s[25104]: Output: mount: /var/SP/rancher/k3s/agent/kubelet/pods/5a3ea7d9-f077-462b-8fcc-645cef659e3a/volume-subpaths/config/test-container/0: mount point is not a directory.
May 05 07:30:47 localhost.localdomain k3s[25104]: W0505 07:30:47.851046      28 mount_helper_common.go:142] Warning: "/var/SP/rancher/k3s/agent/kubelet/pods/5a3ea7d9-f077-462b-8fcc-645cef659e3a/volume-subpaths/config/test-container/0" is not a mountpoint, deleting
May 05 07:30:47 localhost.localdomain k3s[25104]: E0505 07:30:47.851125      28 kubelet_pods.go:237] "Failed to prepare subPath for volumeMount of the container" err=<
May 05 07:30:47 localhost.localdomain k3s[25104]:         error mounting /var/SP/rancher/k3s/agent/kubelet/pods/5a3ea7d9-f077-462b-8fcc-645cef659e3a/volumes/kubernetes.io~configmap/config/..2023_05_03_15_29_28.4065326557/Hello.txt: mount failed: exit status 32
May 05 07:30:47 localhost.localdomain k3s[25104]:         Mounting command: mount
May 05 07:30:47 localhost.localdomain k3s[25104]:         Mounting arguments: --no-canonicalize -o bind,remount /proc/28/fd/153 /var/SP/rancher/k3s/agent/kubelet/pods/5a3ea7d9-f077-462b-8fcc-645cef659e3a/volume-subpaths/config/test-container/0
May 05 07:30:47 localhost.localdomain k3s[25104]:         Output: mount: /var/SP/rancher/k3s/agent/kubelet/pods/5a3ea7d9-f077-462b-8fcc-645cef659e3a/volume-subpaths/config/test-container/0: mount point is not a directory.
May 05 07:30:47 localhost.localdomain k3s[25104]:  > containerName="test-container" volumeMountName="config"
May 05 07:30:47 localhost.localdomain k3s[25104]: E0505 07:30:47.851171      28 kuberuntime_manager.go:872] container &Container{Name:test-container,Image:docker.io/busybox,Command:[sleep 86400],Args:[],WorkingDir:,Ports:[]ContainerPort{},Env:[]EnvVar{},Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{},Claims:[]ResourceClaim{},},VolumeMounts:[]VolumeMount{VolumeMount{Name:config,ReadOnly:false,MountPath:/mnt/Hello.txt,SubPath:Hello.txt,MountPropagation:nil,SubPathExpr:,},VolumeMount{Name:kube-api-access-67qcq,ReadOnly:true,MountPath:/var/run/secrets/kubernetes.io/serviceaccount,SubPath:,MountPropagation:nil,SubPathExpr:,},},LivenessProbe:nil,ReadinessProbe:nil,Lifecycle:nil,TerminationMessagePath:/dev/termination-log,ImagePullPolicy:Always,SecurityContext:nil,Stdin:false,StdinOnce:false,TTY:false,EnvFrom:[]EnvFromSource{},TerminationMessagePolicy:File,VolumeDevices:[]VolumeDevice{},StartupProbe:nil,} start failed in pod test-deployment-5f4ff88f68-s5hwd_default(5a3ea7d9-f077-462b-8fcc-645cef659e3a): CreateContainerConfigError: failed to prepare subPath for volumeMount "config" of container "test-container"
May 05 07:30:47 localhost.localdomain k3s[25104]: E0505 07:30:47.851191      28 pod_workers.go:965] "Error syncing pod, skipping" err="failed to \"StartContainer\" for \"test-container\" with CreateContainerConfigError: \"failed to prepare subPath for volumeMount \\\"config\\\" of container \\\"test-container\\\"\"" pod="default/test-deployment-5f4ff88f68-s5hwd" podUID=5a3ea7d9-f077-462b-8fcc-645cef659e3a
```

The above message in the log is actually misleading:
`Output: mount: /var/SP/rancher/k3s/agent/kubelet/pods/5a3ea7d9-f077-462b-8fcc-645cef659e3a/volume-subpaths/config/test-container/0: mount point is not a directory.`

The mount actually fails with a permission denied. The below patch to [https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/mount-utils/mount_linux.go](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/mount-utils/mount_linux.go) or [https://github.com/kubernetes/mount-utils/blob/master/mount_linux.go](https://github.com/kubernetes/mount-utils/blob/master/mount_linux.go) fixes the issue and creates the expected behaviour:

```diff
diff --git a/mount_linux.go b/mount_linux.go
index f0125fc..e593132 100644
--- a/mount_linux.go
+++ b/mount_linux.go
@@ -35,6 +35,8 @@ import (
 
        "github.com/moby/sys/mountinfo"
 
+       "golang.org/x/sys/unix"
+
        "k8s.io/klog/v2"
        utilexec "k8s.io/utils/exec"
        utilio "k8s.io/utils/io"
@@ -107,6 +109,38 @@ func (mounter *Mounter) hasSystemd() bool {
        return *mounter.withSystemd
 }
 
+// Do a bind mount including the needed remount for applying the bind opts.
+// If the remount fails figure out if the source filesystem has the ro, nodev,
+// noexec or nosuid flag set and try another remount with the found flags.
+func (mounter *Mounter) BindMountSensitive(mounterPath string, mountCmd string, source string, target string, fstype string, bindOpts []string, bindRemountOpts []string, bindRemountOptsSensitive []string, mountFlags []string, systemdMountRequired bool) error {
+       err := mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindOpts, bindRemountOptsSensitive, mountFlags, systemdMountRequired)
+       if err != nil {
+               return err
+       }
+       err = mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindRemountOpts, bindRemountOptsSensitive, mountFlags, systemdMountRequired)
+       if err == nil {
+               return nil
+       }
+       // Check if the source has ro, nodev, noexec, nosuid flag...
+       var s unix.Statfs_t
+       if err := unix.Statfs(source, &s); err != nil {
+               return &os.PathError{Op: "statfs", Path: source, Err: err}
+       }
+       // ... and retry the mount with flags found above.
+       flagMapping := map[int64]string {
+               unix.MS_RDONLY: "ro",
+               unix.MS_NODEV:  "nodev",
+               unix.MS_NOEXEC: "noexec",
+               unix.MS_NOSUID: "nosuid",
+       }
+       for k, v := range flagMapping {
+               if s.Flags & k == k {
+                       bindRemountOpts = append(bindRemountOpts, v)
+               }
+       }
+       return mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindRemountOpts, bindRemountOptsSensitive, mountFlags, systemdMountRequired)
+}
+
 // Mount mounts source to target as fstype with given options. 'source' and 'fstype' must
 // be an empty string in case it's not required, e.g. for remount, or for auto filesystem
 // type, where kernel handles fstype for you. The mount 'options' is a list of options,
@@ -127,11 +161,7 @@ func (mounter *Mounter) MountSensitive(source string, target string, fstype stri
        mounterPath := ""
        bind, bindOpts, bindRemountOpts, bindRemountOptsSensitive := MakeBindOptsSensitive(options, sensitiveOptions)
        if bind {
-               err := mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindOpts, bindRemountOptsSensitive, nil /* mountFlags */, mounter.trySystemd)
-               if err != nil {
-                       return err
-               }
-               return mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindRemountOpts, bindRemountOptsSensitive, nil /* mountFlags */, mounter.trySystemd)
+               return mounter.BindMountSensitive(mounterPath, defaultMountCommand, source, target, fstype, bindOpts, bindRemountOpts, bindRemountOptsSensitive, nil /* mountFlags */, mounter.trySystemd)
        }
        // The list of filesystems that require containerized mounter on GCI image cluster
        fsTypesNeedMounter := map[string]struct{}{
@@ -156,11 +186,7 @@ func (mounter *Mounter) MountSensitiveWithoutSystemdWithMountFlags(source string
        mounterPath := ""
        bind, bindOpts, bindRemountOpts, bindRemountOptsSensitive := MakeBindOptsSensitive(options, sensitiveOptions)
        if bind {
-               err := mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindOpts, bindRemountOptsSensitive, mountFlags, false)
-               if err != nil {
-                       return err
-               }
-               return mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindRemountOpts, bindRemountOptsSensitive, mountFlags, false)
+               return mounter.BindMountSensitive(mounterPath, defaultMountCommand, source, target, fstype, bindOpts, bindRemountOpts, bindRemountOptsSensitive, mountFlags, false)
        }
        // The list of filesystems that require containerized mounter on GCI image cluster
        fsTypesNeedMounter := map[string]struct{}{
```

Similar issues for bind mounts of filesystems with one or more of `ro`, `nodev`, `noexec`, `nosuid` set are fixed in [crun](https://github.com/containers/crun/blob/645b75b04613debb0a3bdf122fdb0b22d835db68/src/libcrun/linux.c#L918-L941) or are under review for [runc](https://github.com/opencontainers/runc/pull/3805).

**Comment by brandond** [comment]
If the issue needs to be patched in core Kubernetes code, I'm curious why you opened the issue here instead of opening an issue and PR against Kubernetes itself.

**Comment by brandond** [comment]
cc @AkihiroSuda

**Comment by rpluem-vf** [comment]
1. Because I wasn't sure if running rootless is something you only support in k3s and hence needs to be patched here locally.
2. I discovered it with k3s and I am not sure if Kubernetes would accept these kind of reports.

**Comment by rpluem-vf** [comment]
But I can open an issue at https://github.com/kubernetes/mount-utils/ if this is better / helps.

**Comment by brandond** [comment]
https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/mount-utils is the correct upstream; `kubernetes/mount-utils` is staged out of the main Kubernetes repo.

You will probably also have to petition for a backport if you want it fixed on anything other than 1.28.

**Comment by AkihiroSuda** [comment]
> BindMountSensitive

This might be more robust
https://github.com/moby/moby/blob/v23.0.5/daemon/oci_linux.go#L430-L460

```go
// Get the set of mount flags that are set on the mount that contains the given
// path and are locked by CL_UNPRIVILEGED. This is necessary to ensure that
// bind-mounting "with options" will not fail with user namespaces, due to
// kernel restrictions that require user namespace mounts to preserve
// CL_UNPRIVILEGED locked flags.
func getUnprivilegedMountFlags(path string) ([]string, error) {
	var statfs unix.Statfs_t
	if err := unix.Statfs(path, &statfs); err != nil {
		return nil, err
	}

	// The set of keys come from https://github.com/torvalds/linux/blob/v4.13/fs/namespace.c#L1034-L1048.
	unprivilegedFlags := map[uint64]string{
		unix.MS_RDONLY:     "ro",
		unix.MS_NODEV:      "nodev",
		unix.MS_NOEXEC:     "noexec",
		unix.MS_NOSUID:     "nosuid",
		unix.MS_NOATIME:    "noatime",
		unix.MS_RELATIME:   "relatime",
		unix.MS_NODIRATIME: "nodiratime",
	}

	var flags []string
	for mask, flag := range unprivilegedFlags {
		if uint64(statfs.Flags)&mask == mask {
			flags = append(flags, flag)
		}
	}

	return flags, nil
}
```

**Comment by rpluem-vf** [comment]
@AkihiroSuda : I guess like the below then, correct?:

```diff
diff --git a/mount_linux.go b/mount_linux.go
index f0125fc..87a8aa1 100644
--- a/mount_linux.go
+++ b/mount_linux.go
@@ -35,6 +35,8 @@ import (
 
 	"github.com/moby/sys/mountinfo"
 
+	"golang.org/x/sys/unix"
+
 	"k8s.io/klog/v2"
 	utilexec "k8s.io/utils/exec"
 	utilio "k8s.io/utils/io"
@@ -107,6 +109,43 @@ func (mounter *Mounter) hasSystemd() bool {
 	return *mounter.withSystemd
 }
 
+// Do a bind mount including the needed remount for applying the bind opts.
+// If the remount fails figure out if the source filesystem has the ro, nodev,
+// noexec, nosuid, noatime, relatime or nodiratime flag set and try another
+// remount with the found flags.
+func (mounter *Mounter) BindMountSensitive(mounterPath string, mountCmd string, source string, target string, fstype string, bindOpts []string, bindRemountOpts []string, bindRemountOptsSensitive []string, mountFlags []string, systemdMountRequired bool) error {
+	err := mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindOpts, bindRemountOptsSensitive, mountFlags, systemdMountRequired)
+	if err != nil {
+		return err
+	}
+	err = mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindRemountOpts, bindRemountOptsSensitive, mountFlags, systemdMountRequired)
+	if err == nil {
+		return nil
+	}
+	// Check if the source has ro, nodev, noexec, nosuid, noatime, relatime,
+	// nodiratime flag...
+	var s unix.Statfs_t
+	if err := unix.Statfs(source, &s); err != nil {
+		return &os.PathError{Op: "statfs", Path: source, Err: err}
+	}
+	// ... and retry the mount with flags found above.
+	flagMapping := map[int64]string {
+		unix.MS_RDONLY:     "ro",
+		unix.MS_NODEV:      "nodev",
+		unix.MS_NOEXEC:     "noexec",
+		unix.MS_NOSUID:     "nosuid",
+		unix.MS_NOATIME:    "noatime",
+		unix.MS_RELATIME:   "relatime",
+		unix.MS_NODIRATIME: "nodiratime",
+	}
+	for k, v := range flagMapping {
+		if s.Flags & k == k {
+			bindRemountOpts = append(bindRemountOpts, v)
+		}
+	}
+	return mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindRemountOpts, bindRemountOptsSensitive, mountFlags, systemdMountRequired)
+}
+
 // Mount mounts source to target as fstype with given options. 'source' and 'fstype' must
 // be an empty string in case it's not required, e.g. for remount, or for auto filesystem
 // type, where kernel handles fstype for you. The mount 'options' is a list of options,
@@ -127,11 +166,7 @@ func (mounter *Mounter) MountSensitive(source string, target string, fstype stri
 	mounterPath := ""
 	bind, bindOpts, bindRemountOpts, bindRemountOptsSensitive := MakeBindOptsSensitive(options, sensitiveOptions)
 	if bind {
-		err := mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindOpts, bindRemountOptsSensitive, nil /* mountFlags */, mounter.trySystemd)
-		if err != nil {
-			return err
-		}
-		return mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindRemountOpts, bindRemountOptsSensitive, nil /* mountFlags */, mounter.trySystemd)
+		return mounter.BindMountSensitive(mounterPath, defaultMountCommand, source, target, fstype, bindOpts, bindRemountOpts, bindRemountOptsSensitive, nil /* mountFlags */, mounter.trySystemd)
 	}
 	// The list of filesystems that require containerized mounter on GCI image cluster
 	fsTypesNeedMounter := map[string]struct{}{
@@ -156,11 +191,7 @@ func (mounter *Mounter) MountSensitiveWithoutSystemdWithMountFlags(source string
 	mounterPath := ""
 	bind, bindOpts, bindRemountOpts, bindRemountOptsSensitive := MakeBindOptsSensitive(options, sensitiveOptions)
 	if bind {
-		err := mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindOpts, bindRemountOptsSensitive, mountFlags, false)
-		if err != nil {
-			return err
-		}
-		return mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindRemountOpts, bindRemountOptsSensitive, mountFlags, false)
+		return mounter.BindMountSensitive(mounterPath, defaultMountCommand, source, target, fstype, bindOpts, bindRemountOpts, bindRemountOptsSensitive, mountFlags, false)
 	}
 	// The list of filesystems that require containerized mounter on GCI image cluster
 	fsTypesNeedMounter := map[string]struct{}{

```

**Comment by rpluem-vf** [comment]
Meanwhile there is an upstream PR: https://github.com/kubernetes/kubernetes/pull/118457 . Unfortunately it is still stuck before initial tests as these need to be approved before they get run.

**Comment by rpluem-vf** [comment]
https://github.com/kubernetes/kubernetes/pull/118457 is now merged and it will be part of 1.31. There are currently no proposals for a backport.

**Comment by brandond** [comment]
Fixed only 10 months later! Sounds about right for Kubernetes.

**Comment by brandond** [comment]
Closing as fixed upstream in v1.31.
